crawl-ref/source/player-stats.cc

```diff
@@ -690,3 +691,132 @@ bool have_stat_zero()
+ *becoming drained if successful.
+ *
+ *Cannot shift below stat zero, and cannot use if drained (checked in
+ *ability.cc)
+*/
+spret_type gnoll_shift_attributes()
+{
+    stat_type source_attribute;
+    stat_type destination_attribute;
+    int keyin;
+#ifdef TOUCH_UI
+    Popup *pop = new Popup("Shift Attributes");
+    MenuEntry *status = new MenuEntry("", MEL_SUBTITLE);
+    pop->push_entry(new MenuEntry(stat_gain_message + " Shift from:", MEL_TITLE));
+    pop->push_entry(status);
+    MenuEntry *me = new MenuEntry("Strength", MEL_ITEM, 0, 'S', false);
+    me->add_tile(tile_def(TILEG_FIGHTING_ON, TEX_GUI));
+    pop->push_entry(me);
+    me = new MenuEntry("Intelligence", MEL_ITEM, 0, 'I', false);
+    me->add_tile(tile_def(TILEG_SPELLCASTING_ON, TEX_GUI));
+    pop->push_entry(me);
+    me = new MenuEntry("Dexterity", MEL_ITEM, 0, 'D', false);
+    me->add_tile(tile_def(TILEG_DODGING_ON, TEX_GUI));
+    pop->push_entry(me);
+    me = new MenuEntry("Cancel", MEL_ITEM, 0, 'X', false);
+    me->add_tile(tile_def(TILEG_PROMPT_NO, TEX_GUI));
+    pop->push_entry(me);
+#else
+    mprf(MSGCH_PROMPT, "Shift from (S)trength, (I)ntelligence, or (D)exterity? ");
+#endif
+    mouse_control mc(MOUSE_MODE_PROMPT);
+#ifdef TOUCH_UI
+    keyin = pop->pop();
+#else
+    keyin = getchm();
+#endif
+    switch (keyin)
+    {
+    case 's':
+    case 'S':
+        source_attribute = STAT_STR;
+        break;
+
+    case 'i':
+    case 'I':
+        source_attribute = STAT_INT;
+        break;
+
+    case 'd':
+    case 'D':
+        source_attribute = STAT_DEX;
+        break;
+
+    case 'X':
+    default:
+        mprf(MSGCH_PROMPT, "Okay, then.");
+        return SPRET_ABORT;
+    }
+
+    if ( (you.stat(source_attribute) - 3) <= 0 )
+    {
+        mprf("You do not have enough %s to transfer!", _stat_name(source_attribute).c_str());
+        return SPRET_ABORT;
+    }
+#ifdef TOUCH_UI
+    Popup *pop = new Popup("Shift Attributes");
+    MenuEntry *status = new MenuEntry("", MEL_SUBTITLE);
+    pop->push_entry(new MenuEntry(stat_gain_message + " Shift to:", MEL_TITLE));
+    pop->push_entry(status);
+    MenuEntry *me = new MenuEntry("Strength", MEL_ITEM, 0, 'S', false);
+    me->add_tile(tile_def(TILEG_FIGHTING_ON, TEX_GUI));
+    pop->push_entry(me);
+    me = new MenuEntry("Intelligence", MEL_ITEM, 0, 'I', false);
+    me->add_tile(tile_def(TILEG_SPELLCASTING_ON, TEX_GUI));
+    pop->push_entry(me);
+    me = new MenuEntry("Dexterity", MEL_ITEM, 0, 'D', false);
+    me->add_tile(tile_def(TILEG_DODGING_ON, TEX_GUI));
+    pop->push_entry(me);
+    me = new MenuEntry("Cancel", MEL_ITEM, 0, 'X', false);
+    me->add_tile(tile_def(TILEG_PROMPT_NO, TEX_GUI));
+    pop->push_entry(me);
+#else
+    mprf(MSGCH_PROMPT, "Shift to (S)trength, (I)ntelligence, or (D)exterity? ");
+#endif
+    mouse_control mc2(MOUSE_MODE_PROMPT);
+#ifdef TOUCH_UI
+    keyin = pop->pop();
+#else
+    keyin = getchm();
+#endif
+
+    switch (keyin)
```

Comment:
Instead of repeating this switch, extract it out to a function.

Reply:
Fixed by moving TOUCH_UI and switch statements for choosing a stat into a single function instead.